### PR TITLE
[NEW] Add new endpoint 'livechat/room.saveInfo' & deprecate 'livechat:saveInfo' meteor method

### PR DESCRIPTION
**Pull request by murtaza98**
## Further comments

Jira: EOC-9 OC-358

PS: this also covers the issue fixed over here -> https://github.com/RocketChat/Rocket.Chat/pull/26747

### CODE REVIEW

File: apps/meteor/app/livechat/server/methods/saveInfo.js [hunk 1]
```diff
@@ -7,6 +7,7 @@ import { LivechatRooms } from '../../../models/server';
 import { callbacks } from '../../../../lib/callbacks';
 import { Livechat } from '../lib/Livechat';
 
+// TODO: Deprecate this method in favor of "livechat/room.saveInfo" endpoint in v6.0.0
```
Comment (MartinSchoeler) [inline apps/meteor/app/livechat/server/methods/saveInfo.js:10]: You can use the JSDoc deprecation notation for this 
https://jsdoc.app/tags-deprecated.html

File: apps/meteor/app/livechat/server/api/rest.js [hunk 2]
```diff
@@ -10,3 +10,4 @@ import './v1/room.js';
 import './v1/videoCall.js';
 import './v1/transfer.js';
 import './v1/contact';
+import './v1/room.ts';
```
Comment (KevLehman) [inline apps/meteor/app/livechat/server/api/rest.js:13]: How complex would be to convert the room.js file to ts?

**Review by KevLehman — DISMISSED** [review]

File: apps/meteor/app/livechat/server/api/rest.js (hunk 2 above)
Comment (murtaza98) [reply apps/meteor/app/livechat/server/api/rest.js:13]: I tried to do that first however it was indeed getting complicated, hence I decided to just create a new ts file now for this particular endpoint & then we can migrate other endpoints one by one in this file on future eng days

File: apps/meteor/app/livechat/server/methods/saveInfo.js (hunk 1 above)
Comment (murtaza98) [reply apps/meteor/app/livechat/server/methods/saveInfo.js]: That was helpful, thanks!

File: apps/meteor/app/livechat/server/api/v1/room.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
+/* eslint-disable no-unreachable */
```
[the remark this answers is not in the document]
Comment (murtaza98) [reply apps/meteor/app/livechat/server/api/v1/room.ts]: Oops! Removed it now

File: apps/meteor/app/livechat/server/methods/saveInfo.js [hunk 4]
```diff
@@ -7,6 +7,7 @@ import { LivechatRooms } from '../../../models/server';
...
 Meteor.methods({
```
[the remark this answers is not in the document]
Comment (murtaza98) [reply apps/meteor/app/livechat/server/methods/saveInfo.js:14]: Yups, good idea :heavy_check_mark:

File: apps/meteor/app/livechat/server/api/v1/room.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+			}
+
+			await Promise.allSettled([Livechat.saveGuest(guestData, this.userId), Livechat.saveRoomInfo(roomData)]);
+
+			callbacks.run('livechat.saveInfo', LivechatRooms.findOneById(roomData._id), {
+				user: this.user,
+				oldRoom: room,
+			});
+
+			return API.v1.success({});
```
Comment (KevLehman) [inline apps/meteor/app/livechat/server/api/v1/room.ts:36]: You can return empty here and change the typing to mark this as returning `void`.

File: packages/rest-typings/src/v1/omnichannel.ts [hunk 6]
```diff
@@ -1082,6 +1164,9 @@ export type OmnichannelEndpoints = {
 	'/v1/livechat/room.forward': {
 		POST: (params: LiveChatRoomForward) => { success: boolean };
 	};
+	'/v1/livechat/room.saveInfo': {
+		POST: (params: LiveChatRoomSaveInfo) => { success: boolean };
```
Comment (KevLehman) [inline packages/rest-typings/src/v1/omnichannel.ts:1168]:
```suggestion
		POST: (params: LiveChatRoomSaveInfo) => void;
```

File: packages/rest-typings/src/v1/omnichannel.ts [hunk 7]
```diff
@@ -305,6 +305,88 @@ const LiveChatRoomForwardSchema = {
...
+				},
+				email: {
+					type: 'string',
+					nullable: true,
+				},
+				phone: {
+					type: 'string',
+					nullable: true,
+				},
+				livechatData: {
```
Comment (KevLehman) [inline packages/rest-typings/src/v1/omnichannel.ts:346]: `livechatData` is an object similar to `[k: string]: string`.

File: packages/rest-typings/src/v1/omnichannel.ts [hunk 8]
```diff
@@ -305,6 +305,88 @@ const LiveChatRoomForwardSchema = {
...
+					nullable: true,
+				},
+				tags: {
+					type: 'array',
+					items: {
+						type: 'string',
+					},
+					nullable: true,
+				},
+				livechatData: {
```
Comment (KevLehman) [inline packages/rest-typings/src/v1/omnichannel.ts:371]: Same here

File: apps/meteor/tests/end-to-end/api/livechat/00-rooms.ts [hunk 9]
```diff
@@ -998,4 +1002,177 @@ describe('LIVECHAT - rooms', function () {
 			expect(body.messages[1]).to.have.property('username', visitor.username);
 		});
 	});
+
+	describe('livechat/room.saveInfo', () => {
```
Comment (KevLehman) [inline apps/meteor/tests/end-to-end/api/livechat/00-rooms.ts:1006]: Can you add more tests for the cases of guestData/roomData not being sent, for sending them as other but objects, as sending invalid values on the keys, etc.

**Review by KevLehman — DISMISSED** [review]

File: apps/meteor/tests/end-to-end/api/livechat/00-rooms.ts (hunk 9 above)
Comment (murtaza98) [reply apps/meteor/tests/end-to-end/api/livechat/00-rooms.ts:1102]: For sure :smile: Added a couple of more tests to validate the payloads

**Review by sampaiodiego — APPROVED** [review]

**Review by KevLehman — DISMISSED** [review]

**Review by KevLehman — DISMISSED** [review]

File: apps/meteor/app/livechat/server/methods/saveInfo.js [hunk 10]
```diff
@@ -6,9 +6,16 @@ import { hasPermission } from '../../../authorization';
 import { LivechatRooms } from '../../../models/server';
 import { callbacks } from '../../../../lib/callbacks';
 import { Livechat } from '../lib/Livechat';
+import { methodDeprecationLogger } from '../../../lib/server/lib/deprecationWarningLogger';
 
+/**
+ * @deprecated since version 6.0.0
```
Comment (sampaiodiego) [inline apps/meteor/app/livechat/server/methods/saveInfo.js:12]: it is actually deprecated NOW but will be **removed** on version 6.0.0 😬

**Review by sampaiodiego — APPROVED** [review]

**Comment by sampaiodiego** [comment]
should we flag PRs with new endpoints as `[NEW]` ? 🤔

File: apps/meteor/app/livechat/server/methods/saveInfo.js (hunk 10 above)
Comment (murtaza98) [reply apps/meteor/app/livechat/server/methods/saveInfo.js]: Oops.. that's a mistake. Let me fix that :+1:

**Review by KevLehman — APPROVED** [review]

**Comment by murtaza98** [comment]
Hi @sampaiodiego @MartinSchoeler Lost your approval here due to new commits. Do you mind checking this PR again and approving it if all looks good? :hugs:

**Review by sampaiodiego — APPROVED** [review]

**Comment by sampaiodiego** [comment]
do it

**Review by MartinSchoeler — APPROVED** [review]
Everything seems to be OK on the FE

**Pull request merged**